sycl/include/CL/sycl/queue.hpp

```diff
@@ -754,285 +754,238 @@ class __SYCL_EXPORT queue {
                      _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
     _CODELOCARG(&CodeLoc);
     return submit(
         [&](handler &CGH) {
           CGH.depends_on(DepEvent);
-          CGH.template parallel_for<KernelName, KernelType>(
-              NumWorkItems, WorkItemOffset, KernelFunc);
+          CGH.template parallel_for<KernelName>(Range, WorkItemOffset,
+                                                KernelFunc);
         },
         CodeLoc);
   }
 
   /// parallel_for version with a kernel represented as a lambda + range and
   /// offset that specify global size and global offset correspondingly.
   ///
-  /// \param NumWorkItems is a range that specifies the work space of the kernel
+  /// \param Range specifies the global work space of the kernel
   /// \param WorkItemOffset specifies the offset for each work item id
   /// \param DepEvents is a vector of events that specifies the kernel
   /// dependencies
   /// \param KernelFunc is the Kernel functor or lambda
   /// \param CodeLoc contains the code location of user code
   template <typename KernelName = detail::auto_name, typename KernelType,
             int Dims>
-  event parallel_for(range<Dims> NumWorkItems, id<Dims> WorkItemOffset,
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
+  event parallel_for(range<Dims> Range, id<Dims> WorkItemOffset,
                      const std::vector<event> &DepEvents,
                      _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
     _CODELOCARG(&CodeLoc);
     return submit(
         [&](handler &CGH) {
           CGH.depends_on(DepEvents);
-          CGH.template parallel_for<KernelName, KernelType>(
-              NumWorkItems, WorkItemOffset, KernelFunc);
+          CGH.template parallel_for<KernelName>(Range, WorkItemOffset,
+                                                KernelFunc);
         },
         CodeLoc);
   }
 
   /// parallel_for version with a kernel represented as a lambda + nd_range that
   /// specifies global, local sizes and offset.
   ///
-  /// \param ExecutionRange is a range that specifies the work space of the
-  /// kernel
-  /// \param KernelFunc is the Kernel functor or lambda
-  /// \param CodeLoc contains the code location of user code
-  template <typename KernelName = detail::auto_name, typename KernelType,
-            int Dims>
-  event parallel_for(nd_range<Dims> ExecutionRange,
-                     _KERNELFUNCPARAM(KernelFunc) _CODELOCPARAM(&CodeLoc)) {
-    _CODELOCARG(&CodeLoc);
+  /// \param Range specifies the global and local work spaces of the kernel
+  /// \param Rest acts as-if: "ReductionTypes&&... Reductions,
+  /// const KernelType &KernelFunc".
+  template <typename KernelName = detail::auto_name, int Dims,
+            typename... RestT>
+  std::enable_if_t<
+      ext::oneapi::detail::AreAllButLastReductions<RestT...>::value, event>
+  parallel_for(nd_range<Dims> Range, RestT &&...Rest) {
+    const detail::code_location CodeLoc = {};
```

Comment:
We are sending a dummy code location for now, so can you please add a comment here that says "Actual code location needs to be captured from KernelInfo object" ?

Reply:
I added the comment in this additional commit: fd064e221e8e692430f21b3e9d9295e418afa199
In the process of adding the comment, I noticed some redundancy in parallel_for_impl accepting 'offset' (deprecated shortcuts)